File: packages/python/aillium_schemas/remote_assist.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Remote Assist Cloud Contracts
# ---------------------------------------------------------------------------


class EndpointLifecycleState(str, Enum):
    PENDING_ENROLLMENT = "pending_enrollment"
    ENROLLED = "enrolled"
    ACTIVE = "active"
    MAINTENANCE = "maintenance"
    SUSPENDED = "suspended"
    QUARANTINED = "quarantined"
    RETIRED = "retired"
    ARCHIVED = "archived"
    DELETED = "deleted"


class EndpointHealthStatus(str, Enum):
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    CRITICAL = "critical"
    UNKNOWN = "unknown"


class AssistanceRequestCategory(str, Enum):
    DEVICE_IS_SLOW = "device_is_slow"
    CANNOT_SIGN_IN = "cannot_sign_in"
    VPN_OR_NETWORK_ISSUE = "vpn_or_network_issue"
    APP_NOT_WORKING = "app_not_working"
    EMAIL_OR_CALENDAR_ISSUE = "email_or_calendar_issue"
    PRINTER_OR_PERIPHERAL_ISSUE = "printer_or_peripheral_issue"
    SOFTWARE_INSTALL_OR_UPDATE = "software_install_or_update"
    SECURITY_CONCERN = "security_concern"
    SOMETHING_ELSE = "something_else"


class AssistanceRequestPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class AssistanceRequestStatus(str, Enum):
    SUBMITTED = "submitted"
    TRIAGED = "triaged"
    AWAITING_APPROVAL = "awaiting_approval"
    SCHEDULED = "scheduled"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    ESCALATED = "escalated"
    REJECTED = "rejected"


class SupportMode(str, Enum):
    DIAGNOSTICS_ONLY = "diagnostics_only"
    LIVE_HELP = "live_help"
    ALLOW_UNATTENDED_FIX = "allow_unattended_fix"
    APPROVAL_REQUIRED_BEFORE_ANY_ACTION = "approval_required_before_any_action"


class SessionType(str, Enum):
    REMOTE_DESKTOP = "remote_desktop"
    TERMINAL = "terminal"
    DIAGNOSTICS = "diagnostics"
    AI_ASSISTED = "ai_assisted"


class SessionStatus(str, Enum):
    PENDING = "pending"
    ACTIVE = "active"
    PAUSED = "paused"
    ENDED = "ended"
    FAILED = "failed"


class SessionApprovalState(str, Enum):
    NOT_REQUIRED = "not_required"
    PENDING = "pending"
    APPROVED = "approved"
    DENIED = "denied"


class ActionRiskLevel(str, Enum):
    READ_ONLY = "read_only"
    LOW_RISK_WRITE = "low_risk_write"
    HIGH_RISK_WRITE = "high_risk_write"
    RESTRICTED = "restricted"


class RAEvidenceType(str, Enum):
    SCREENSHOT = "screenshot"
    LOG_CAPTURE = "log_capture"
    COMMAND_OUTPUT = "command_output"
    SCREEN_RECORDING_METADATA = "screen_recording_metadata"
    CONFIG_SNAPSHOT = "config_snapshot"


class ReportResultStatus(str, Enum):
    RESOLVED = "resolved"
    PARTIALLY_RESOLVED = "partially_resolved"
    UNRESOLVED = "unresolved"
    ESCALATED = "escalated"


class AuditResult(str, Enum):
    SUCCESS = "success"
    FAILURE = "failure"
    DENIED = "denied"
    ERROR = "error"


class RAPolicyType(str, Enum):
    UNATTENDED_ACCESS = "unattended_access"
    COMMAND_ALLOWLIST = "command_allowlist"
    COMMAND_BLOCKLIST = "command_blocklist"
    FILE_TRANSFER_RULES = "file_transfer_rules"
    AI_ASSIST_PERMISSIONS = "ai_assist_permissions"
    UI_AUTOMATION_RULES = "ui_automation_rules"
    APPROVAL_REQUIREMENTS = "approval_requirements"
    SESSION_RETENTION = "session_retention"
    EVIDENCE_RETENTION = "evidence_retention"
    REDACTION_RULES = "redaction_rules"


class RAPolicyStatus(str, Enum):
    ACTIVE = "active"
    DRAFT = "draft"
    DISABLED = "disabled"
    ARCHIVED = "archived"


# Entity models


class RAEndpoint(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.endpoint"]
    id: str
    tenant_id: str
    workspace_id: str | None = None
    hostname: str = Field(min_length=1)
    serial_number: str | None = None
    manufacturer: str | None = None
    model: str | None = None
    os_name: str | None = None
    os_version: str | None = None
    agent_version: str | None = None
    assigned_user_id: str | None = None
    lifecycle_state: EndpointLifecycleState
    health_status: EndpointHealthStatus
    last_seen_at: datetime | None = None
    tags: list[str] = Field(default_factory=list)
    site: str | None = None
    mesh_device_id: str | None = None
    created_at: datetime
    updated_at: datetime


class RAEndpointLifecycleEvent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.endpoint_lifecycle_event"]
    id: str
    tenant_id: str
    endpoint_id: str
    from_state: EndpointLifecycleState
    to_state: EndpointLifecycleState
    changed_by: str = Field(min_length=1)
    reason: str | None = None
    created_at: datetime


class RAEndpointHealthSnapshot(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.endpoint_health_snapshot"]
    id: str
    tenant_id: str
    endpoint_id: str
    cpu_usage: float | None = None
    memory_usage: float | None = None
    disk_usage: float | None = None
    network_status: str | None = None
    agent_status: str | None = None
    captured_at: datetime


class RAAssistanceRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.assistance_request"]
    id: str
    tenant_id: str
    workspace_id: str | None = None
    requester_user_id: str
    endpoint_id: str | None = None
    category: AssistanceRequestCategory
    priority: AssistanceRequestPriority
    description: str = Field(min_length=1)
    support_mode: SupportMode
    consent_mode: str | None = None
    status: AssistanceRequestStatus
    assigned_technician_id: str | None = None
    submitted_at: datetime
    updated_at: datetime
    resolved_at: datetime | None = None


class RARequestAttachment(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.request_attachment"]
    id: str
    tenant_id: str
    request_id: str
    file_name: str = Field(min_length=1)
    file_type: str = Field(min_length=1)
    storage_uri: str = Field(min_length=1)
    created_at: datetime


class RARequestConsent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.request_consent"]
    id: str
    tenant_id: str
    request_id: str
    consent_type: str = Field(min_length=1)
    granted: bool
    granted_by: str = Field(min_length=1)
    granted_at: datetime


class RASupportSession(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.support_session"]
    id: str
    tenant_id: str
    endpoint_id: str
    request_id: str | None = None
    session_type: SessionType
    technician_id: str
    status: SessionStatus
    approval_state: SessionApprovalState
    started_at: datetime
    ended_at: datetime | None = None
    duration_seconds: int | None = None


class RASessionAction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.session_action"]
    id: str
    tenant_id: str
    session_id: str
    action_type: str = Field(min_length=1)
    tool_used: str | None = None
    actor_id: str = Field(min_length=1)
    command_or_action_summary: str | None = None
    risk_level: ActionRiskLevel
    approval_reference: str | None = None
    result: str | None = None
    created_at: datetime


class RASessionEvidence(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.session_evidence"]
    id: str
    tenant_id: str
    session_id: str
    evidence_type: RAEvidenceType
    storage_uri: str = Field(min_length=1)
    hash: str | None = None
    created_at: datetime


class RASessionReport(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.session_report"]
    id: str
    tenant_id: str
    session_id: str
    request_id: str | None = None
    summary: str = Field(min_length=1)
    actions_taken: Any | None = None
    validation_result: str | None = None
    result_status: ReportResultStatus
    technician_notes: str | None = None
    ai_summary: str | None = None
    created_at: datetime


class RAAuditEvent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.audit_event"]
    id: str
    tenant_id: str
    actor_id: str = Field(min_length=1)
    actor_role: str = Field(min_length=1)
    target_type: str = Field(min_length=1)
    target_id: str | None = None
    event_type: str = Field(min_length=1)
    event_summary: str = Field(min_length=1)
    result: AuditResult
    risk_level: str | None = None
    ip_address: str | None = None
    metadata_json: Any | None = None
    created_at: datetime


class RAPolicy(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.policy"]
    id: str
    tenant_id: str
    name: str = Field(min_length=1)
    policy_type: RAPolicyType
    policy_json: Any
    status: RAPolicyStatus
    created_by: str = Field(min_length=1)
    created_at: datetime
    updated_at: datetime


class RATenantEntitlement(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.tenant_entitlement"]
    id: str
    tenant_id: str
    feature_key: str = Field(min_length=1)
    enabled: bool
    limit_value: int | None = None
    current_usage: int = 0
    billing_source: str | None = None
    created_at: datetime
    updated_at: datetime


class RABillingUsageRecord(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_type: Literal["remote_assist.billing_usage_record"]
    id: str
    tenant_id: str
    feature_key: str = Field(min_length=1)
    usage_unit: str = Field(min_length=1)
    usage_value: float
    period_start: datetime
    period_end: datetime
    source_reference: str | None = None
    created_at: datetime

File: packages/typescript/src/remote-assist.ts
import { z } from "zod";

// ---------------------------------------------------------------------------
// Remote Assist Cloud Contracts
// ---------------------------------------------------------------------------

// Enums

export const EndpointLifecycleStateValues = [
  "pending_enrollment",
  "enrolled",
  "active",
  "maintenance",
  "suspended",
  "quarantined",
  "retired",
  "archived",
  "deleted",
] as const;
export type EndpointLifecycleState = (typeof EndpointLifecycleStateValues)[number];
export const EndpointLifecycleStateSchema = z.enum(EndpointLifecycleStateValues);

export const EndpointHealthStatusValues = ["healthy", "degraded", "critical", "unknown"] as const;
export type EndpointHealthStatus = (typeof EndpointHealthStatusValues)[number];
export const EndpointHealthStatusSchema = z.enum(EndpointHealthStatusValues);

export const AssistanceRequestCategoryValues = [
  "device_is_slow",
  "cannot_sign_in",
  "vpn_or_network_issue",
  "app_not_working",
  "email_or_calendar_issue",
  "printer_or_peripheral_issue",
  "software_install_or_update",
  "security_concern",
  "something_else",
] as const;
export type AssistanceRequestCategory = (typeof AssistanceRequestCategoryValues)[number];
export const AssistanceRequestCategorySchema = z.enum(AssistanceRequestCategoryValues);

export const AssistanceRequestPriorityValues = ["low", "medium", "high", "urgent"] as const;
export type AssistanceRequestPriority = (typeof AssistanceRequestPriorityValues)[number];
export const AssistanceRequestPrioritySchema = z.enum(AssistanceRequestPriorityValues);

export const AssistanceRequestStatusValues = [
  "submitted",
  "triaged",
  "awaiting_approval",
  "scheduled",
  "in_progress",
  "resolved",
  "escalated",
  "rejected",
] as const;
export type AssistanceRequestStatus = (typeof AssistanceRequestStatusValues)[number];
export const AssistanceRequestStatusSchema = z.enum(AssistanceRequestStatusValues);

export const SupportModeValues = [
  "diagnostics_only",
  "live_help",
  "allow_unattended_fix",
  "approval_required_before_any_action",
] as const;
export type SupportMode = (typeof SupportModeValues)[number];
export const SupportModeSchema = z.enum(SupportModeValues);

export const SessionTypeValues = ["remote_desktop", "terminal", "diagnostics", "ai_assisted"] as const;
export type SessionType = (typeof SessionTypeValues)[number];
export const SessionTypeSchema = z.enum(SessionTypeValues);

export const SessionStatusValues = ["pending", "active", "paused", "ended", "failed"] as const;
export type SessionStatus = (typeof SessionStatusValues)[number];
export const SessionStatusSchema = z.enum(SessionStatusValues);

export const SessionApprovalStateValues = ["not_required", "pending", "approved", "denied"] as const;
export type SessionApprovalState = (typeof SessionApprovalStateValues)[number];
export const SessionApprovalStateSchema = z.enum(SessionApprovalStateValues);

export const ActionRiskLevelValues = ["read_only", "low_risk_write", "high_risk_write", "restricted"] as const;
export type ActionRiskLevel = (typeof ActionRiskLevelValues)[number];
export const ActionRiskLevelSchema = z.enum(ActionRiskLevelValues);

export const EvidenceTypeValues = [
  "screenshot",
  "log_capture",
  "command_output",
  "screen_recording_metadata",
  "config_snapshot",
] as const;
export type EvidenceType = (typeof EvidenceTypeValues)[number];
export const EvidenceTypeSchema = z.enum(EvidenceTypeValues);

export const ReportResultStatusValues = ["resolved", "partially_resolved", "unresolved", "escalated"] as const;
export type ReportResultStatus = (typeof ReportResultStatusValues)[number];
export const ReportResultStatusSchema = z.enum(ReportResultStatusValues);

export const AuditResultValues = ["success", "failure", "denied", "error"] as const;
export type AuditResult = (typeof AuditResultValues)[number];
export const AuditResultSchema = z.enum(AuditResultValues);

export const PolicyTypeValues = [
  "unattended_access",
  "command_allowlist",
  "command_blocklist",
  "file_transfer_rules",
  "ai_assist_permissions",
  "ui_automation_rules",
  "approval_requirements",
  "session_retention",
  "evidence_retention",
  "redaction_rules",
] as const;
export type PolicyType = (typeof PolicyTypeValues)[number];
export const PolicyTypeSchema = z.enum(PolicyTypeValues);

export const PolicyStatusValues = ["active", "draft", "disabled", "archived"] as const;
export type PolicyStatus = (typeof PolicyStatusValues)[number];
export const PolicyStatusSchema = z.enum(PolicyStatusValues);

// Entity schemas

export const RAEndpointSchema = z.object({
  contract_type: z.literal("remote_assist.endpoint"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  workspace_id: z.string().uuid().nullable().optional(),
  hostname: z.string().min(1),
  serial_number: z.string().nullable().optional(),
  manufacturer: z.string().nullable().optional(),
  model: z.string().nullable().optional(),
  os_name: z.string().nullable().optional(),
  os_version: z.string().nullable().optional(),
  agent_version: z.string().nullable().optional(),
  assigned_user_id: z.string().uuid().nullable().optional(),
  lifecycle_state: EndpointLifecycleStateSchema,
  health_status: EndpointHealthStatusSchema,
  last_seen_at: z.string().datetime().nullable().optional(),
  tags: z.array(z.string()).default([]),
  site: z.string().nullable().optional(),
  mesh_device_id: z.string().nullable().optional(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});
export type RAEndpoint = z.infer<typeof RAEndpointSchema>;

export const RAEndpointLifecycleEventSchema = z.object({
  contract_type: z.literal("remote_assist.endpoint_lifecycle_event"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  endpoint_id: z.string().uuid(),
  from_state: EndpointLifecycleStateSchema,
  to_state: EndpointLifecycleStateSchema,
  changed_by: z.string().min(1),
  reason: z.string().nullable().optional(),
  created_at: z.string().datetime(),
});
export type RAEndpointLifecycleEvent = z.infer<typeof RAEndpointLifecycleEventSchema>;

export const RAEndpointHealthSnapshotSchema = z.object({
  contract_type: z.literal("remote_assist.endpoint_health_snapshot"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  endpoint_id: z.string().uuid(),
  cpu_usage: z.number().nullable().optional(),
  memory_usage: z.number().nullable().optional(),
  disk_usage: z.number().nullable().optional(),
  network_status: z.string().nullable().optional(),
  agent_status: z.string().nullable().optional(),
  captured_at: z.string().datetime(),
});
export type RAEndpointHealthSnapshot = z.infer<typeof RAEndpointHealthSnapshotSchema>;

export const RAAssistanceRequestSchema = z.object({
  contract_type: z.literal("remote_assist.assistance_request"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  workspace_id: z.string().uuid().nullable().optional(),
  requester_user_id: z.string().uuid(),
  endpoint_id: z.string().uuid().nullable().optional(),
  category: AssistanceRequestCategorySchema,
  priority: AssistanceRequestPrioritySchema,
  description: z.string().min(1),
  support_mode: SupportModeSchema,
  consent_mode: z.string().nullable().optional(),
  status: AssistanceRequestStatusSchema,
  assigned_technician_id: z.string().uuid().nullable().optional(),
  submitted_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  resolved_at: z.string().datetime().nullable().optional(),
});
export type RAAssistanceRequest = z.infer<typeof RAAssistanceRequestSchema>;

export const RARequestAttachmentSchema = z.object({
  contract_type: z.literal("remote_assist.request_attachment"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  request_id: z.string().uuid(),
  file_name: z.string().min(1),
  file_type: z.string().min(1),
  storage_uri: z.string().min(1),
  created_at: z.string().datetime(),
});
export type RARequestAttachment = z.infer<typeof RARequestAttachmentSchema>;

export const RARequestConsentSchema = z.object({
  contract_type: z.literal("remote_assist.request_consent"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  request_id: z.string().uuid(),
  consent_type: z.string().min(1),
  granted: z.boolean(),
  granted_by: z.string().min(1),
  granted_at: z.string().datetime(),
});
export type RARequestConsent = z.infer<typeof RARequestConsentSchema>;

export const RASupportSessionSchema = z.object({
  contract_type: z.literal("remote_assist.support_session"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  endpoint_id: z.string().uuid(),
  request_id: z.string().uuid().nullable().optional(),
  session_type: SessionTypeSchema,
  technician_id: z.string().uuid(),
  status: SessionStatusSchema,
  approval_state: SessionApprovalStateSchema,
  started_at: z.string().datetime(),
  ended_at: z.string().datetime().nullable().optional(),
  duration_seconds: z.number().int().nullable().optional(),
});
export type RASupportSession = z.infer<typeof RASupportSessionSchema>;

export const RASessionActionSchema = z.object({
  contract_type: z.literal("remote_assist.session_action"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  session_id: z.string().uuid(),
  action_type: z.string().min(1),
  tool_used: z.string().nullable().optional(),
  actor_id: z.string().min(1),
  command_or_action_summary: z.string().nullable().optional(),
  risk_level: ActionRiskLevelSchema,
  approval_reference: z.string().nullable().optional(),
  result: z.string().nullable().optional(),
  created_at: z.string().datetime(),
});
export type RASessionAction = z.infer<typeof RASessionActionSchema>;

export const RASessionEvidenceSchema = z.object({
  contract_type: z.literal("remote_assist.session_evidence"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  session_id: z.string().uuid(),
  evidence_type: EvidenceTypeSchema,
  storage_uri: z.string().min(1),
  hash: z.string().nullable().optional(),
  created_at: z.string().datetime(),
});
export type RASessionEvidence = z.infer<typeof RASessionEvidenceSchema>;

export const RASessionReportSchema = z.object({
  contract_type: z.literal("remote_assist.session_report"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  session_id: z.string().uuid(),
  request_id: z.string().uuid().nullable().optional(),
  summary: z.string().min(1),
  actions_taken: z.unknown().optional(),
  validation_result: z.string().nullable().optional(),
  result_status: ReportResultStatusSchema,
  technician_notes: z.string().nullable().optional(),
  ai_summary: z.string().nullable().optional(),
  created_at: z.string().datetime(),
});
export type RASessionReport = z.infer<typeof RASessionReportSchema>;

export const RAAuditEventSchema = z.object({
  contract_type: z.literal("remote_assist.audit_event"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  actor_id: z.string().min(1),
  actor_role: z.string().min(1),
  target_type: z.string().min(1),
  target_id: z.string().nullable().optional(),
  event_type: z.string().min(1),
  event_summary: z.string().min(1),
  result: AuditResultSchema,
  risk_level: z.string().nullable().optional(),
  ip_address: z.string().nullable().optional(),
  metadata_json: z.unknown().optional(),
  created_at: z.string().datetime(),
});
export type RAAuditEvent = z.infer<typeof RAAuditEventSchema>;

export const RAPolicySchema = z.object({
  contract_type: z.literal("remote_assist.policy"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  name: z.string().min(1),
  policy_type: PolicyTypeSchema,
  policy_json: z.unknown(),
  status: PolicyStatusSchema,
  created_by: z.string().min(1),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});
export type RAPolicy = z.infer<typeof RAPolicySchema>;

export const RATenantEntitlementSchema = z.object({
  contract_type: z.literal("remote_assist.tenant_entitlement"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  feature_key: z.string().min(1),
  enabled: z.boolean(),
  limit_value: z.number().int().nullable().optional(),
  current_usage: z.number().int().default(0),
  billing_source: z.string().nullable().optional(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});
export type RATenantEntitlement = z.infer<typeof RATenantEntitlementSchema>;

export const RABillingUsageRecordSchema = z.object({
  contract_type: z.literal("remote_assist.billing_usage_record"),
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  feature_key: z.string().min(1),
  usage_unit: z.string().min(1),
  usage_value: z.number(),
  period_start: z.string().datetime(),
  period_end: z.string().datetime(),
  source_reference: z.string().nullable().optional(),
  created_at: z.string().datetime(),
});
export type RABillingUsageRecord = z.infer<typeof RABillingUsageRecordSchema>;
